unit-tests-java8/src/test/java/org/eclipse/collections/test/bag/mutable/sorted/MutableBagIterableTestCase.java

```diff
@@ -37,6 +38,17 @@ default void MutableBagIterable_addOccurrences()
         assertThrows(
                 IllegalArgumentException.class,
                 () -> mutableBag.addOccurrences(4, -1));
+        assertEquals(Bags.immutable.with(1, 1, 1, 2, 2, 3, 3, 3, 4, 4, 4, 4), mutableBag);
+        assertEquals(Bags.immutable.with(1, 1, 1, 2, 2, 3, 3, 3, 4, 4, 4, 4), mutableBag);
+
+        assertThrows(
+                ArithmeticException.class,
+                () -> mutableBag.addOccurrences(3, Integer.MAX_VALUE - 1));
+        assertEquals(Bags.immutable.with(1, 1, 1, 2, 2, 3, 3, 3, 4, 4, 4, 4), mutableBag);
```

Comment:
With the changes here, we can no longer overflow the number of occurrences for a single key.

Reply:
`MutableBag.addOccurances` does not specify in its javadoc that it will throw `ArithmeticException`. And that probably applies to everywhere else those functions are involved.

Reply:
`strictAddOccurrences` is fine, then the next question would be whether to call it internally. For example, in HashBag.collect():

```java
    @Override
    public <V, R extends Collection<V>> R collect(Function<? super T, ? extends V> function, R target)
    {
        if (target instanceof MutableBagIterable<?>)
        {
            MutableBagIterable<V> targetBag = (MutableBagIterable<V>) target;

            this.forEachWithOccurrences((each, occurrences) -> targetBag.addOccurrences(function.valueOf(each), occurrences));
        }
        else
        {
            // ...
        }
        return target;
    }
```

Or add strictCollect()? Or just stop here?

Reply:
An alternative approach is to have a `StrictMutableBag` wrapper (like synchronized). Then the question of "where to stop" goes away.

Reply:
> It's hard to figure out where to stop, as I don't understand why it was started in the first place. Is there a real world use case that triggered this?

So, I don't remember. I wrote the first version of this commit 6 years ago. I was talking to @donraab about having so much work in progress that I had almost given up on contributing and decided to try to land some of the smaller changes.

I think that the overflow has always bothered me and so I've always avoided the `addOccurrences` method. At my work, there are no usages of Bag in the company. In my side projects I've used `addOccurrences` [exactly once](https://github.com/motlin/klass/blob/edbd1efa8f68f91d80c91f88d877bb7adf736ef4/klass-reladomo/klass-reladomo-sample-data/src/main/java/cool/klass/reladomo/sample/data/AbstractDataTypePropertyVisitor.java#L196), and it was with the occurrences hard-coded to `1`. I just used the method so that I could increment and get the new count in a single method call.

> An alternative approach is to have a StrictMutableBag wrapper (like synchronized). Then the question of "where to stop" goes away.

> It occurred to me that implementing that wrapper would be easier if MutableBag has a strictAddOccurances method, so maybe a combination of the two works best.

I like these ideas and will try them out. I'll move this code review to a Draft, I don't plan on landing it in a form anything like it is now.

Reply:
Thanks for jumping in on this @mohrezaei. I like your idea of having a `StrictMutableBag` if this functionality is desired. Another idea I have been pondering for a while is having a `LargeBag` which uses `long` instead of `int` for the counts. I haven't needed it, so haven't bothered. The backing Map is still limited by `int` for it's size. Java `Stream` counts into `Map<K, Long>`, which allows for the larger counts, while still suffering from the `int` backing size.